tensorboard/data/server/server.rs

```diff
@@ -312,15 +507,72 @@ mod tests {
                 plugin_name: "scalars".to_string(),
                 ..Default::default()
             }),
+            run_tag_filter: Some(data::RunTagFilter {
+                runs: Some(data::RunFilter {
+                    names: vec!["train".to_string(), "nonexistent".to_string()],
+                    ..Default::default()
+                }),
+                tags: None,
+                ..Default::default()
+            }),
             downsample: Some(data::Downsample {
                 num_points: 1000,
                 ..Default::default()
             }),
             ..Default::default()
         });
+
         let res = handler.read_scalars(req).await.unwrap().into_inner();
+
         assert_eq!(res.runs.len(), 1);
-        assert_eq!(res.runs[0].tags.len(), 1);
-        // fake data; don't bother checking the contents
+        let map = unroll_map(
+            res.runs,
+            |x| (x.run_name, x.tags),
+            |y| (y.tag_name, y.data.unwrap()),
+        );
+
+        let train_run = &map[&Run("train".to_string())];
+        assert_eq!(train_run.len(), 1);
+        let xent_data = &train_run[&Tag("xent".to_string())];
+        assert_eq!(xent_data.step, vec![0, 1, 2]);
+        assert_eq!(xent_data.wall_time, vec![1235.0, 1236.0, 1237.0]);
+        assert_eq!(xent_data.value, vec![0.5, 0.25, 0.125]);
+    }
+
+    #[tokio::test]
+    async fn test_read_scalars_needs_downsample() {
+        let handler = sample_handler();
+        let req = Request::new(data::ReadScalarsRequest {
+            experiment_id: "123".to_string(),
+            plugin_filter: Some(data::PluginFilter {
+                plugin_name: "scalars".to_string(),
+                ..Default::default()
+            }),
+            downsample: None,
+            ..Default::default()
+        });
+        let res = handler.read_scalars(req).await.unwrap_err();
+        match (res.code(), res.message()) {
+            (Code::InvalidArgument, msg) if msg.contains("downsample") => (),
+            other => panic!("{:?}", other),
+        };
+    }
+
+    #[tokio::test]
+    async fn test_read_scalars_downsample_zero_okay() {
+        let handler = sample_handler();
+        let req = Request::new(data::ReadScalarsRequest {
+            experiment_id: "123".to_string(),
+            plugin_filter: Some(data::PluginFilter {
+                plugin_name: "scalars".to_string(),
+                ..Default::default()
+            }),
+            downsample: Some(data::Downsample {
+                num_points: 0,
+                ..Default::default()
+            }),
+            ..Default::default()
+        });
+        handler.read_scalars(req).await.unwrap();
```

Comment:
Do we not want to actually assert anything on the response?  Seems like it might not hurt to check that there are indeed no points.